skyportal/facility_apis/sedm.py

```diff
@@ -0,0 +1,214 @@
+            photometry_payload[p.filter[-1]] = {
+                'jd': p.mjd + 2_400_000.5,
+                'mag': p.mag,
+                'obsdate': p.iso.date().isoformat(),
+            }
+
+    rtype = request.payload['observation_type']
+    if rtype == 'IFU':
+        filters = ''
+        followup = 'IFU'
+    elif rtype == '3-shot (gri)':
+        filters = 'g,r,i'
+        followup = ''
+    elif rtype == '4-shot (ugri)':
+        filters = 'u,g,r,i'
+        followup = ''
+    elif rtype == '4-shot+IFU':
+        filters = 'u,g,r,i'
+        followup = 'IFU'
+    elif rtype == '3-shot+IFU':
+        filters = 'g,r,i'
+        followup = 'IFU'
+    elif rtype == "Mix 'n Match":
+        choices = request.payload['observation_choices']
+        hasspec = 'IFU' in choices
+        followup = 'IFU' if hasspec else ''
+        if hasspec:
+            choices.remove('IFU')
+        filters = ','.join(choices)
+    else:
+        raise ValueError('Cannot coerce payload into SEDM format.')
+
+    payload = {
+        'Filters': filters,
+        'Followup': followup,
+        'email': request.requester.contact_email,
+        'enddate': request.payload['end_date'],
+        'startdate': request.payload['start_date'],
+        'prior_photometry': photometry_payload,
+        'priority': request.payload['priority'],
+        'programname': request.allocation.group.name,
+        'requestid': request.id,
+        'sourceid': request.obj_id,
+        'sourcename': request.obj_id,
+        'status': method_value,
+        'username': request.requester.username,
+    }
+
+    return payload
+
+
+class SEDMAPI(FollowUpAPI):
+    """SkyPortal interface to the Spectral Energy Distribution machine (SEDM)."""
+
+    @staticmethod
+    def submit(request):
+        """Submit a follow-up request to SEDM."""
+        from ..models import DBSession, FollowupRequestHTTPRequest
+
+        payload = convert_request_to_sedm(request, method_value='new')
+        content = json.dumps(payload)
+        r = requests.post(
+            cfg['app.sedm_endpoint'], files={'jsonfile': ('jsonfile', content)},
+        )
+
+        if r.status_code == 200:
+            request.status = 'submitted'
+        else:
+            request.status = f'rejected: {r.content}'
+
+        message = FollowupRequestHTTPRequest(
+            content=dump.dump_all(r).decode('utf-8'),
+            origin='skyportal',
+            request=request,
+        )
+        DBSession().add(message)
+        DBSession().add(request)
+        DBSession().commit()
+
+    @staticmethod
+    def delete(request):
+        """Delete a follow-up request from SEDM queue."""
+        from ..models import DBSession, FollowupRequest
+
+        payload = convert_request_to_sedm(request, method_value='delete')
+        content = json.dumps(payload)
+        r = requests.post(
+            cfg['app.sedm_endpoint'], files={'jsonfile': ('jsonfile', content)},
+        )
+
+        r.raise_for_status()
```

Comment:
We do not want unhandled exceptions in our backend.

Reply:
This is handled in the `FollowupRequestHandler`

Reply:
Ah, sorry, thought for a moment there this was a handler.